ipa-core/src/protocol/context/mod.rs

```diff
@@ -312,7 +316,35 @@ impl<'a, B: ShardBinding> Context for Base<'a, B> {
 
 /// Context for MPC circuits that can operate on multiple shards. Provides access to shard information
 /// via [`ShardConfiguration`] trait.
-pub trait ShardedContext: Context + ShardConfiguration {}
+pub trait ShardedContext: Context + ShardConfiguration {
+    /// Requests data to be received from all shards that are registered in the system.
+    /// Shards that don't have any data to send, must explicitly open and close the send channel.
+    fn recv_from_shards<M: Message>(
+        &self,
+    ) -> impl Stream<Item = (ShardIndex, Result<M, crate::error::Error>)> + Send {
+        stream::select_all(
+            self.peer_shards()
+                .map(|origin| self.shard_recv_channel(origin).map(move |v| (origin, v))),
+        )
+    }
+
+    /// Picks a shard according to the value obtained from sampling PRSS shared with the given helper.
+    fn pick_shard(&self, record_id: RecordId, direction: Direction) -> ShardIndex {
+        // FIXME: update PRSS trait to compute only left or right part
+        let (l, r): (u128, u128) = self.prss().generate(record_id);
+        let shard_index = u32::try_from(
+            match direction {
+                Direction::Left => l,
+                Direction::Right => r,
+            } % u128::from(self.shard_count()),
+        )
+        .expect("Number of shards should not exceed u32 capacity");
+
+        ShardIndex::from(shard_index)
+    }
+}
+
+impl<C: Context + ShardConfiguration> ShardedContext for C {}
 
 impl ShardConfiguration for Base<'_, Sharded> {
```

Comment:
This isn't directly about this PR, but if we're settling on a "protocols should be written against upgraded contexts" policy (see discussion in #1021), then maybe `ShardConfiguration` shouldn't be implemented for the base context.

Reply:
I merged this PR and then I saw this comment - for some reason it wasn't showing up on the "Changes" tab. I could be wrong but I think we don't expose `Base` context directly - it is always wrapped into a semi-honest or malicious version

Reply:
Yeah, I was mixing things up when I made this comment. By "base context" I meant `semi_honest::Context`, which is not the same as `struct Base`.